### ISSUE TITLE
Feature: Secure & Scalable rollouts (and rollbacks)

### ISSUE THREAD

**Issue by azeemshaikh38** [ossf/scorecard#399]
**Is your feature request related to a problem? Please describe.**
Today, various binaries included in Scorecard need to be built and pushed manually to GCS. What's worse is that, someone with authorization has the capability to push a locally modified binary to production GCS. We need safer and more automated rollout/rollback procedure. 

**Describe the solution you'd like**
There could multiple "right" solutions here, so I'll describe this in terms of levels, each of which builds on top of the previous one. We can start simple and move to higher (and more complex) levels:

Level 1: Build and push manually.
  - Provide scripts which can help users with authorization to push the right binaries to the right GCS images.

Level 2: Build and push manually, but no local changes allowed. Only code which is at main can be pushed.
  - Basically, have some kind of checksum to verify that user is pushing code which matches code present in main (at HEAD or otherwise). If not, fail.

Level 3: Automate build and push from GitHub.
  - Update all existing unit and integration tests to be non-hermetic and non-flaky. Today, most tests depend on external production services and can be flaky. 
  - Integrate GitHub and GCS such that GitHub can build and push a selected commit/version to GCS.
  - Provide a way for users to manually select a commit/version that GitHub will push to GCS. This allows for easy rollbacks if we ever need one.

Level 4: Canaries and automated rollback. 

@naveensrinivasan assigning this to you for now, let me know what you think. We can use this issue to discuss potential solutions.

**Comment by azeemshaikh38** [comment]
@oliverchang @ laurentsimon fyi.

**Comment by naveensrinivasan** [comment]
All the main branch changes are already in gcr

Instead of push it is pull into gcr 

The Cloud build watches for changes to the GitHub and does container build and pushes it to the gcr.

**Comment by inferno-chromium** [comment]
> All the main branch changes are already in gcr
> 
> Instead of push it is pull into gcr
> 
> The Cloud build watches for changes to the GitHub and does container build and pushes it to the gcr.

Do we need all this code then - https://github.com/ossf/scorecard/blob/main/.github/workflows/docker.yaml maybe time to remove ?

**Comment by naveensrinivasan** [comment]
We need this code to keep an container image in GitHub.

we can decide to open up GCR and then remove this.

**Comment by inferno-chromium** [comment]
> We need this code to keep an container image in GitHub.
> 
> we can decide to open up GCR and then remove this.

yes we should, make gcr public and then remove this. @azeemshaikh38 fyi

**Comment by naveensrinivasan** [comment]
Ok. I will do a PR for that.

**Comment by azeemshaikh38** [comment]
Thanks Naveen! 2 questions:

1. With the pull model, how do rollbacks works? I assume we will have to rollback main/HEAD, is that correct?
2. Can we also make an update such that no manual pushes are allowed to GCR? This might be too strict for now, but we can look into figuring out a better solution later.

**Comment by naveensrinivasan** [comment]
> Thanks Naveen! 2 questions:
> 
> 1. With the pull model, how do rollbacks works? I assume we will have to rollback main/HEAD, is that correct?

Yes! 

> 2. Can we also make an update such that no manual pushes are allowed to GCR? This might be too strict for now, but we can look into figuring out a better solution later.

The pushes are based on RBAC on GCP.  That is where it can be controlled. There isn't another to control this.

**Comment by naveensrinivasan** [comment]
Closing this issue as we have rollback.
